src/validation/validate_wasm.rs

```diff
@@ -0,0 +1,237 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use anyhow::Result;
+use colored::Colorize;
+use impl_serde::serialize as serde_hex;
+use parity_wasm::elements::Module;
+
+/// Marker inserted by the ink! codegen for an error which can't
+/// be checked at compile time.
+const INK_ENFORCE_ERR: &str = "__ink_enforce_error_";
+
+/// Errors which may occur when forwarding a call is not allowed.
+///
+/// We insert markers for these errors in the generated contract code.
+/// This is necessary since we can't check these errors at compile time
+/// of the contract.
+/// `cargo-contract` checks the contract code for these error markers
+/// when building a contract and fails if it finds markers.
+#[derive(codec::Encode, codec::Decode)]
+pub enum EnforcedErrors {
+    /// The below error represents calling a `&mut self` message in a context that
+    /// only allows for `&self` messages. This may happen under certain circumstances
+    /// when ink! trait implementations are involved with long-hand calling notation.
+    #[codec(index = 1)]
+    CannotCallTraitMessage {
+        /// The trait that defines the called message.
+        trait_ident: String,
+        /// The name of the called message.
+        message_ident: String,
+        /// The selector of the called message.
+        message_selector: [u8; 4],
+        /// Is `true` if the `self` receiver of the ink! message is `&mut self`.
+        message_mut: bool,
+    },
+    /// The below error represents calling a constructor in a context that does
+    /// not allow calling it. This may happen when the constructor defined in a
+    /// trait is cross-called in another contract.
+    /// This is not allowed since the contract to which a call is forwarded must
+    /// already exist at the point when the call to it is made.
+    #[codec(index = 2)]
+    CannotCallTraitConstructor {
+        /// The trait that defines the called constructor.
+        trait_ident: String,
+        /// The name of the called constructor.
+        constructor_ident: String,
+        /// The selector of the called constructor.
+        constructor_selector: [u8; 4],
+    },
+}
+
+/// Validates the import section in the Wasm.
+///
+/// The checks currently fall into two categories:
+/// - Known bugs for which we want to recommend a solution.
+/// - Markers inserted by the ink! codegen for errors which can't be checked at compile time.
+pub fn validate_import_section(module: &Module) -> Result<()> {
+    let imports = module
+        .import_section()
+        .expect("import section must exist")
+        .entries()
+        .iter();
+    let original_imports_len = imports.len();
+    let mut errs = Vec::new();
+
+    let filtered_imports = imports.filter(|section| {
```

Comment:
This whole block would be (for me) more readable if it was either a pure loop which pushed to `errs` and increments `import_len` or purely functional by aggregating the errors. Something like this *might* work

```rust
let result = 
        imports
            .map(|import| {
                let field = section.field();
                if field.contains("panic") {
                    Err(anyhow::anyhow!(String::from(
                        "An unexpected panic function import was found in the contract Wasm.\n\
                        This typically goes back to a known bug in the Rust compiler:\n\
                        https://github.com/rust-lang/rust/issues/78744\n\n\
                        As a workaround try to insert `overflow-checks = false` into your `Cargo.toml`.\n\
                        This will disable safe math operations, but unfortunately we are currently not \n\
                        aware of a better workaround until the bug in the compiler is fixed.",
                    )))
                } else if field.starts_with(INK_ENFORCE_ERR) {
                    Err(anyhow::anyhow!(parse_linker_error(field)))
                } else if !is_import_allowed(field) {
                    Err(anyhow::anyhow!("TODO import not allowed"))
                } else {
                    Ok(())
                }
            })
            .collect::<Result<(), Vec<String>>>();
    
    result.map_err(|errs| {
        anyhow::anyhow!(
            "Validation of the Wasm failed.\n\n\n{}",
            errs.into_iter()
                .map(|err| format!("{} {}", "ERROR:".to_string().bold(), err))
                .collect::<Vec<String>>()
                .join("\n\n\n")
        )
    })
```

Reply:
Actually on second thoughts this won't work because the Result `FromIterator` only collects for the `Ok` variant.